src/cmd/services/m3coordinator/downsample/downsampler.go

```diff
@@ -0,0 +1,756 @@
+	}
+
+	ruleSetOpts := rules.NewOptions().
+		SetTagsFilterOptions(tagsFilterOptions).
+		SetNewRollupIDFn(newRollupIDFn).
+		SetIsRollupIDFn(isRollupIDFn).
+		SetAggregationTypesOptions(aggTypeOpts)
+
+	opts := matcher.NewOptions().
+		SetClockOptions(clockOpts).
+		SetInstrumentOptions(instrumentOpts).
+		SetRuleSetOptions(ruleSetOpts).
+		SetKVStore(rulesStore)
+
+	cacheOpts := cache.NewOptions().
+		SetClockOptions(clockOpts).
+		SetInstrumentOptions(instrumentOpts.
+			SetMetricsScope(instrumentOpts.MetricsScope().SubScope("matcher-cache")))
+
+	cache := cache.NewCache(cacheOpts)
+
+	matcher, err := matcher.NewMatcher(cache, opts)
+	if err != nil {
+		return newAggregatorResult{}, err
+	}
+
+	aggregatorOpts := aggregator.NewOptions().
+		SetClockOptions(clockOpts).
+		SetInstrumentOptions(instrumentOpts).
+		SetAggregationTypesOptions(aggTypeOpts).
+		SetMetricPrefix(nil).
+		SetCounterPrefix(nil).
+		SetGaugePrefix(nil).
+		SetTimerPrefix(nil)
+
+	shardSet := make([]shard.Shard, numShards)
+	for i := 0; i < numShards; i++ {
+		shardSet[i] = shard.NewShard(uint32(i)).
+			SetState(shard.Initializing).
+			SetCutoverNanos(0).
+			SetCutoffNanos(math.MaxInt64)
+	}
+	shards := shard.NewShards(shardSet)
+	instance := placement.NewInstance().
+		SetID(instanceID).
+		SetShards(shards).
+		SetShardSetID(shardSetID)
+	localPlacement := placement.NewPlacement().
+		SetInstances([]placement.Instance{instance}).
+		SetShards(shards.AllIDs())
+	stagedPlacement := placement.NewStagedPlacement().
+		SetPlacements([]placement.Placement{localPlacement})
+	stagedPlacementProto, err := stagedPlacement.Proto()
+	if err != nil {
+		return newAggregatorResult{}, err
+	}
+
+	placementStore := mem.NewStore()
+	_, err = placementStore.SetIfNotExists(placementKVKey, stagedPlacementProto)
+	if err != nil {
+		return newAggregatorResult{}, err
+	}
+
+	placementWatcherOpts := placement.NewStagedPlacementWatcherOptions().
+		SetStagedPlacementKey(placementKVKey).
+		SetStagedPlacementStore(placementStore)
+	placementWatcher := placement.NewStagedPlacementWatcher(placementWatcherOpts)
+	placementManagerOpts := aggregator.NewPlacementManagerOptions().
+		SetInstanceID(instanceID).
+		SetStagedPlacementWatcher(placementWatcher)
+	placementManager := aggregator.NewPlacementManager(placementManagerOpts)
+	aggregatorOpts = aggregatorOpts.SetPlacementManager(placementManager)
+
+	// Set up flush times manager.
+	flushTimesManagerOpts := aggregator.NewFlushTimesManagerOptions().
+		SetFlushTimesStore(placementStore)
+	flushTimesManager := aggregator.NewFlushTimesManager(flushTimesManagerOpts)
+	aggregatorOpts = aggregatorOpts.SetFlushTimesManager(flushTimesManager)
+
+	// Set up election manager.
+	leaderValue := instanceID
+	campaignOpts, err := services.NewCampaignOptions()
+	if err != nil {
+		return newAggregatorResult{}, err
+	}
+
+	campaignOpts = campaignOpts.SetLeaderValue(leaderValue)
+	electionCluster := integration.NewClusterV3(nil, &integration.ClusterConfig{
```

Comment:
:\ so the coordinator also starts an embedded single-node etcd cluster? Why not pass in the etcd cluster externally so we can use the embedded KV cluster in M3DB nodes if we choose to?

Reply:
So we don't have an external etcd cluster.  What actually needs to happen here is we need to remove the dependency of a real etcd cluster from the leader service, it should just take a struct that can fulfill some things and abstract the underlying implementation (which happens to be etcd).